### PR TITLE
Fix invalid compile time errors when expected type is map and data projection is false

### LINKED ISSUE #7329: ParserAsType gives error when we disable the data projection for map types
### Description

_No response_

### Steps to Reproduce

The issue can be reproduced by executing(bal run) the below ballerina code,
note: this error occurs for every data type with map, ex: map\<int>, map\<string>, map<byte[]>, etc. 

[1] Ballerina code
```ballerina
import ballerina/data.jsondata;
public function main() returns error? {
    json a = {id: "chamil", values: {a: 2, b: 45, c: {x: "mnb", y: "uio"}}};
    map<json> _ = check jsondata:parseAsType(a, options = {allowDataProjection: false});
}
```
[2] Error
```
error: undefined field 'id'
```

### Version

0.3.0-20241028-143400-903c253

### Environment Details (with versions)

_No response_

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes https://github.com/ballerina-platform/ballerina-library/issues/7329

### CODE REVIEW

File: ballerina/tests/from_json_test.bal [hunk 1]
```diff
@@ -1637,3 +1637,23 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
...
+function testMapAsExpectedTypeWithJsonSource() returns error? {
+    json jsonValue = {id: "chamil", values: {a: 2, b: 45, c: {x: "mnb", y: "uio"}}};
+
+    map<json> mapValue = check parseAsType(jsonValue, options = {allowDataProjection: false});
+    test:assertEquals(mapValue, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
+
+    record{|json...;|} recValue = check parseAsType(jsonValue, options = {allowDataProjection: false});
+    test:assertEquals(recValue, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
+
+    map<json> mapValue2 = check parseAsType(jsonValue, options = {allowDataProjection: {}});
```
Comment (lnash94) [inline ballerina/tests/from_json_test.bal:1770]: Shall we add the test value with nillabe as well `map<json>?` ?

Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal:1770]: Added in 3e80cae882625b4498d4602241ee5d4d344c5245

**Comment by lnash94** [comment]
You will have more test cases by referring to this commit https://github.com/ballerina-platform/module-ballerina-http/pull/2204/commits/17d630405f72a1731046db7d532ec4b5a7891fd3

File: ballerina/tests/from_json_test.bal [hunk 2]
```diff
@@ -1752,3 +1752,29 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
     ReadonlyFieldsRec27 r7 = check parseAsType(user);
     test:assertEquals(r7, {testId: 4012, testTaxNo: "N/A", testName: "John Doe"});
 }
+
+@test:Config
+function testMapAsExpectedTypeWithJsonSource() returns error? {
+    json jsonValue = {id: "chamil", values: {a: 2, b: 45, c: {x: "mnb", y: "uio"}}};
+
+    map<json> mapValue = check parseAsType(jsonValue, options = {allowDataProjection: false});
+    test:assertEquals(mapValue, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
```
Comment (MaryamZi) [inline ballerina/tests/from_json_test.bal]:
```suggestion
    test:assertEquals(mapValue, {"id": "chamil", "values": {"a": 2, "b": 45, "c": {"x": "mnb", "y": "uio"}}});
```

File: ballerina/tests/from_json_test.bal [hunk 3]
```diff
@@ -1752,3 +1752,29 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
...
+    test:assertEquals(recValue2, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
+
+    map<anydata> mapValue3 = check parseAsType(jsonValue, options = {allowDataProjection: {}});
+    test:assertEquals(mapValue3, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
+
+    map<json>? mapValue4 = check parseAsType(jsonValue, options = {allowDataProjection: false});
+    test:assertEquals(mapValue4, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
+
+    record{|json?...;|} recValue3 = check parseAsType(jsonValue, options = {allowDataProjection: false});
+    test:assertEquals(recValue3, {"id":"chamil","values":{"a":2,"b":45,"c":{"x":"mnb","y":"uio"}}});
```
Comment (MaryamZi) [inline ballerina/tests/from_json_test.bal]: Can we add tests for more types as mentioned in the issue? Not just `json`.

Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal]: Update in 907da11a2710ca5da519003e7920cad082df7733

File: ballerina/tests/from_json_test.bal (hunk 2 above)
Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal]: Update in 907da11a2710ca5da519003e7920cad082df7733

File: ballerina/tests/from_json_test.bal [hunk 4]
```diff
@@ -1752,3 +1752,43 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
...
+    test:assertEquals(mapValue5, {id: 1, age: 23});
+
+    map<string> mapValue6 = check parseAsType(jsonValue3, options = {allowDataProjection: false});
+    test:assertEquals(mapValue6, {name: "abc", address: "N/A"});
+
+    record{|int...;|} recValue4 = check parseAsType(jsonValue2, options = {allowDataProjection: false});
+    test:assertEquals(recValue4, {id: 1, age: 23});
+
+    record{|string...;|} recValue5 = check parseAsType(jsonValue3, options = {allowDataProjection: false});
+    test:assertEquals(recValue5, {name: "abc", address: "N/A"});
```
Comment (MaryamZi) [inline ballerina/tests/from_json_test.bal:1803]: This doesn't really assert if the value has been parsed as the type, right? We can test updating it in an invalid way.

Comment (MaryamZi) [reply ballerina/tests/from_json_test.bal:1803]: And/or use the is check.

File: ballerina/tests/from_json_test.bal [hunk 5]
```diff
@@ -1752,3 +1752,43 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
...
+    map<int> mapValue5 = check parseAsType(jsonValue2, options = {allowDataProjection: false});
+    test:assertEquals(mapValue5, {id: 1, age: 23});
+
+    map<string> mapValue6 = check parseAsType(jsonValue3, options = {allowDataProjection: false});
+    test:assertEquals(mapValue6, {name: "abc", address: "N/A"});
+
+    record{|int...;|} recValue4 = check parseAsType(jsonValue2, options = {allowDataProjection: false});
+    test:assertEquals(recValue4, {id: 1, age: 23});
+
+    record{|string...;|} recValue5 = check parseAsType(jsonValue3, options = {allowDataProjection: false});
```
Comment (MaryamZi) [inline ballerina/tests/from_json_test.bal:1802]: Would add with union rest descriptors too.

Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal:1802]: Fixed in aa624d1aeee7d01c956c3259756b68615f155d1f

File: ballerina/tests/from_json_test.bal (hunk 4 above)
Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal:1803]: Fixed in aa624d1aeee7d01c956c3259756b68615f155d1fF

File: ballerina/tests/from_json_test.bal [hunk 6]
```diff
@@ -1752,3 +1752,62 @@ function testReadonlyFieldsWithNameAnnotation() returns error? {
...
+    test:assertEquals(recValue4, {id: 1, age: 23});
+    test:assertTrue(recValue4 is record{|int...;|});
+
+    record{|string...;|} recValue5 = check parseAsType(jsonValue3, options = {allowDataProjection: false});
+    test:assertEquals(recValue5, {name: "abc", address: "N/A"});
+    test:assertTrue(recValue5 is record{|string...;|});
+
+    map<record{|json...;|}|string|int> mapValue7 = check parseAsType(jsonValue, options = {allowDataProjection: {}});
+    test:assertEquals(mapValue7, {"id": "chamil", "values": {"a":2, "b": 45, "c": {"x": "mnb", "y": "uio"}}});
+    test:assertTrue(mapValue7 is map<anydata>);
```
Comment (MaryamZi) [inline ballerina/tests/from_json_test.bal]: What is the expectation with this check?

Comment (MaryamZi) [reply ballerina/tests/from_json_test.bal]: What are we testing?

Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal]: I think the only way to achieve this task, https://github.com/ballerina-platform/module-ballerina-data.jsondata/pull/49#discussion_r1829223509 is this.

Comment (MaryamZi) [reply ballerina/tests/from_json_test.bal]: This doesn't check that, does it? anydata allows way more values than the target map's member type, so you have to use an exact check - `is map<record{|json...;|}|string|int>`.

Comment (SasinduDilshara) [reply ballerina/tests/from_json_test.bal]: Update in 8e98e64d863c7bb9e4a9b942b252e922707aca14

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**